taichi/backends/opengl/opengl_api.cpp

```diff
@@ -503,12 +413,20 @@ struct CompiledKernel::Impl {
        const std::string &kernel_source_code,
        std::unique_ptr<ParallelSize> ps_)
       : kernel_name(kernel_name_), ps(std::move(ps_)) {
-    source =
-        kernel_source_code +
-        fmt::format(
-            "layout(local_size_x = {}, local_size_y = 1, local_size_z = 1) in;",
-            ps->get_threads_per_block());
-    display_kernel_info(kernel_name_, source);
+    if (ps->grid_dim > opengl_max_grid_dim)
+      ps->grid_dim = opengl_max_grid_dim;
+    if (ps->block_dim > opengl_max_block_dim)
+      ps->block_dim = opengl_max_block_dim;
+
+    size_t needle = kernel_source_code.find("precision highp float;\n");
+    TI_ASSERT(needle != std::string::npos);
+    source = kernel_source_code.substr(0, needle) +
+             fmt::format(
+                 "layout(local_size_x = {}, local_size_y = 1, local_size_z = "
+                 "1) in;\n",
+                 ps->block_dim) +
+             kernel_source_code.substr(needle);
```

Comment:
```suggestion
    size_t layout_pos = kernel_source_code.find("precision highp float;\n");
    TI_ASSERT(layout_pos != std::string::npos);
    source = kernel_source_code.substr(0, layout_pos) +
             fmt::format(
                 "layout(local_size_x = {}, local_size_y = 1, local_size_z = "
                 "1) in;\n",
                 ps->block_dim) +
             kernel_source_code.substr(layout_pos);
```
`layout_pos` sounds like a clearer name compared to `needle`.